pkg/commands/waiter.go

```diff
@@ -0,0 +1,183 @@
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/dynamic"
+	watchTools "k8s.io/client-go/tools/watch"
+	"knative.dev/pkg/apis"
+	"knative.dev/pkg/apis/duck"
+	duckv1 "knative.dev/pkg/apis/duck/v1"
+	"knative.dev/pkg/kmeta"
+)
+
+type ResourceWaiter interface {
+	Wait(runtime.Object) error
+	BuilderWait(runtime.Object, int64, int64) error
+}
+
+func NewResourceWaiter(dc dynamic.Interface) ResourceWaiter {
+	return NewWaiter(dc)
+}
+
+type Waiter struct {
+	dynamicClient dynamic.Interface
+}
+
+func NewWaiter(dc dynamic.Interface) *Waiter {
+	return &Waiter{dynamicClient: dc}
+}
+
+func (i *Waiter) Wait(ob runtime.Object) error {
+	m, ok := ob.(kmeta.OwnerRefable)
+	if !ok {
+		return errors.New("unexpected type")
+	}
+	return i.wait(ob, hasResolved(m.GetObjectMeta().GetGeneration()))
+}
+
+// Used to ensure the builder has resolved to the expected stack and store (importing)
+func (i *Waiter) BuilderWait(ob runtime.Object, expectedStoreGeneration, expectedStackGeneration int64) error {
+	m, ok := ob.(kmeta.OwnerRefable)
+	if !ok {
+		return errors.New("unexpected type")
+	}
+	return i.wait(ob, builderHasResolved(m.GetObjectMeta().GetGeneration(), expectedStoreGeneration, expectedStackGeneration))
+}
+
+func (i *Waiter) wait(ob runtime.Object, cf watchTools.ConditionFunc) error {
+	e := &watch.Event{Object: ob}
+	done, err := cf(*e)
+	if err != nil {
+		return err
+	}
+	if !done {
+		m, ok := ob.(kmeta.OwnerRefable)
+		if !ok {
+			return errors.New("unexpected type")
+		}
+		name := m.GetObjectMeta().GetName()
+		gvr, _ := meta.UnsafeGuessKindToResource(m.GetGroupVersionKind())
+		rv := m.GetObjectMeta().GetResourceVersion()
+		w := watchOne{name: name, gvr: gvr, dynamicClient: i.dynamicClient}
+		e, err = watchTools.Until(context.Background(), rv, w, filterErrors(cf))
+		if err != nil {
+			return err
+		}
+	}
+
+	kr, err := eventToDuck(e)
+	if err != nil {
+		return err
+	}
+
+	if condition := kr.Status.GetCondition(apis.ConditionReady); condition.IsFalse() {
+		if condition.Message != "" {
+			return errors.Errorf("%v %q not ready: %v", kr.Kind, kr.Name, condition.Message)
+		}
+
+		return errors.Errorf("%v %q not ready", kr.Kind, kr.Name)
+	}
+	return nil
+}
+
+func eventToDuck(e *watch.Event) (*duckv1.KResource, error) {
+	u := &unstructured.Unstructured{}
+	var err error
+	u.Object, err = runtime.DefaultUnstructuredConverter.ToUnstructured(e.Object)
+	if err != nil {
+		return nil, err
+	}
+
+	kr := &duckv1.KResource{}
+	if err := duck.FromUnstructured(u, kr); err != nil {
+		return nil, err
+	}
+	return kr, nil
+}
+
+func hasResolved(generation int64) func(event watch.Event) (bool, error) {
+	return func(event watch.Event) (bool, error) {
+		kr, err := eventToDuck(&event)
+		if err != nil {
+			return false, err
+		}
+
+		if kr.Status.ObservedGeneration < generation {
+			return false, nil // still waiting on update
+		}
+
+		if kr.Status.GetCondition(apis.ConditionReady).IsUnknown() {
```

Comment:
I think this makes sense. Is there a time we expect this behavior. Generation observed but, unknown?

Reply:
good question! this was maybe a false assumption I made.